File: 0344/two_pointers.cpp
#include <gtest/gtest.h>

#include <cstdlib>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  void reverseString(vector<char>& s) {
    size_t left = 0, right = s.size();
    while (left < right) {
      char tmp = s[left];
      s[left] = s[right-1];
      s[right-1] = tmp;
      left++;
      right--;
    }
  }
};

void check(const vector<char>& expected, const vector<char>& got) {
  ASSERT_EQ(expected.size(), got.size());
  for (size_t i = 0; i < expected.size(); ++i) {
    EXPECT_EQ(expected[i], got[i]);
  }
}

int main() {
  Solution sol;
  vector<char> s, expected;

  s = {'h','e','l','l','o'};
  expected = {'o','l','l','e','h'};
  sol.reverseString(s);
  check(expected, s);

  exit(EXIT_SUCCESS);
}

File: 2938/two_pointers.py
from typing import *

class Solution:
    def minimumSteps(self, s: str) -> int:
        left, right = 0, len(s) - 1
        moves = 0
        while True:
            while left < len(s) and s[left] == "0":
                left += 1
            while 0 <= right and s[right] == "1":
                right -= 1
            if left >= right:
                break
            moves += right - left
            left += 1
            right -= 1
        return moves


if __name__ == '__main__':
    sol = Solution()

    s, expected = "101", 1
    assert sol.minimumSteps(s) == expected

    s, expected = "100", 2
    assert sol.minimumSteps(s) == expected

    s, expected = "0111", 0
    assert sol.minimumSteps(s) == expected
